Documentation/design-docs/object-stack-allocation.md

```diff
@@ -0,0 +1,190 @@
+**Pros:**
+* The jit can analyze callee's code (subject to some restrictions, e.g., when running under profiler)
+  since there are no versioning considerations at jit time.
+* The optimization will apply to any msil code regardless of the language compiler or msil post-processing tools.
+* The jit already has IR that's suitable for escape analysis.
+
+**Cons:**
+* The jit analyzes methods top-down, i.e., callers before callees (when inlining), which doesn't fit well with the stack allocation optimization.
+* Full interprocedural analysis is too expensive for the jit, even at high tiering levels.
+
+Possible approaches to interprocedural analysis in the jit:
+* We can run escape analysis concurrently with inlining and analyze callee's parameters for escaping while inspecting
+inline candidates. The results of such analysis can be cached.
+* We can adjust inlining heuristics to give more weight to candidates whose parameters have references to potentially
+stack-allocated objects. Inlining such methods may result in additional benefits if the jit can promote fields of the
+stack-allocated objects.
+* For higher-tier jit the order of method processing may be closer to bottom-up, i.e., callees before callers. That may
+  help with running stack allocation optimization.
+
+### Analysis in ngen/crossgen
+**Pros:**
+* ngen/crossgen can afford to spend more time for escape analysis.
+* The jit already has IR that's suitable for escape analysis.
+
+**Cons:**
+* crossgen in Ready-To-Run mode is not running on generic methods that cross assembly boundaries.
+* crossgen in Ready-To-Run mode is not allowed to analyze code of methods from other assemblies.
+* newobj in Ready-To-Run mode is more abstract and introducing a hard dependence on ref class size and layout may interfere
+with version reseliency.
+
+### Analysis in ILLink
+**Pros:**
+* ILLInk can afford to spend more time for escape analysis.
+* For self-contained apps, ILLink has access to all of application's code and can do full interprocedural analysis.
+* ILLink is already a part of System.Private.CoreLib and CoreFX build toolchain so the assemblies built there can benefit
+from this.
+
+**Cons:**
+* The implementation will only benefit customers that have ILLink in their toolchain
+* ILLink operated on a view of metadata and raw msil instructions, it currently doesn't have a call graph or IR representation
+suitable for escape analysis.
+
+The results of escape analysis in the linker may be communicated to the jit by injecting an intrinsic call right before or after
+newobj for the object that was determined to be non-escaping. Note that assemblies may lose verifiability with this approach.
+An alternative is to annotate parameters with escape information so that the annotations can be verified by the jit with
+local analysis.
+
+## Other restrictions on stack allocations
+
+* Objects with finalizers can't be stack-allocated since they always escape to the finalizer queue. 
```

Comment:
What about objects that have a finalizer, but whose finalization is always suppressed?

Though I don't know what is the likelihood of determining whether finalization is indeed suppressed, since it might require inlining of `Dispose` (which is the method that commonly calls `SuppressFinalize`).

Reply:
Or could the finalizer just be called synchronously similiar to `using (obj)`

Reply:
I think at least for the initial version it's ok to be conservative here and not stack allocate objects with finalizers without worrying about whether they are suppressed.